src/plugins/data_views/common/data_views/_pattern_cache.ts

```diff
@@ -10,30 +10,50 @@ import { DataView } from './data_view';
 
 export interface DataViewCache {
   get: (id: string) => Promise<DataView> | undefined;
+  getByHash: (id: string) => DataView | undefined;
   set: (id: string, value: Promise<DataView>) => Promise<DataView>;
   clear: (id: string) => void;
   clearAll: () => void;
 }
 
 export function createDataViewCache(): DataViewCache {
-  const vals: Record<string, Promise<DataView>> = {};
+  const vals: Map<string, Promise<DataView>> = new Map();
+  const valsByHash: Map<string, DataView> = new Map();
   const cache: DataViewCache = {
     get: (id: string) => {
-      return vals[id];
+      return vals.get(id);
+    },
+    getByHash: (hash: string) => {
+      return valsByHash.get(hash);
     },
     set: (id: string, prom: Promise<DataView>) => {
-      vals[id] = prom;
+      vals.set(id, prom);
+      prom
+        .then((dv) => {
+          valsByHash.set(dv.getSpecHash(), dv);
+        })
+        .catch(() => {
+          // no reason to keep a failed promise
+          vals.delete(id);
```

Comment:
moving this code from consumers, which shouldn't care about such implementation detail IMO